Client/Frontend/Browser/Handlers/BraveTalkScriptHandler.swift

```diff
@@ -8,75 +8,120 @@ import WebKit
     self.rewards = rewards
+    self.launchNativeBraveTalk = launchNativeBraveTalk
 
     tab.rewardsEnabledCallback = { [weak self] success in
-      self?.callback(result: success)
+      self?.rewardsEnabledReplyHandler?(success, nil)
     }
   }
 
   static func name() -> String { "BraveTalkHelper" }
 
   func scriptMessageHandlerName() -> String? { "BraveTalkHelper_\(UserScriptManager.messageHandlerTokenString)" }
 
+  private struct Payload: Decodable {
+    enum Kind: Decodable {
+      case braveRequestAdsEnabled
+      case launchNativeBraveTalk(String)
+    }
+    var kind: Kind
+    var securityToken: String
+    
+    enum CodingKeys: String, CodingKey {
+      case kind
+      case url
+      case securityToken = "securitytoken"
+    }
+    
+    init(from decoder: Decoder) throws {
+      enum RawKindKey: String, Decodable {
+        case braveRequestAdsEnabled
+        case launchNativeBraveTalk
+      }
+      let container = try decoder.container(keyedBy: CodingKeys.self)
+      let kind = try container.decode(RawKindKey.self, forKey: .kind)
+      self.securityToken = try container.decode(String.self, forKey: .securityToken)
+      switch kind {
+      case .launchNativeBraveTalk:
+        let url = try container.decode(String.self, forKey: .url)
+        self.kind = .launchNativeBraveTalk(url)
+      case .braveRequestAdsEnabled:
+        self.kind = .braveRequestAdsEnabled
+      }
+    }
+  }
+  
   func userContentController(
     _ userContentController: WKUserContentController,
     didReceiveScriptMessage message: WKScriptMessage,
-    replyHandler: (Any?, String?) -> Void
+    replyHandler: @escaping (Any?, String?) -> Void
   ) {
-    defer { replyHandler(nil, nil) }
     let allowedHosts = DomainUserScript.braveTalkHelper.associatedDomains
 
     guard let requestHost = message.frameInfo.request.url?.host,
       allowedHosts.contains(requestHost),
       message.frameInfo.isMainFrame
     else {
-      log.error("Backup search request called from disallowed host")
+      log.error("Brave Talk request called from disallowed host")
+      return
+    }
+    
+    guard let json = try? JSONSerialization.data(withJSONObject: message.body, options: []),
+          let payload = try? JSONDecoder().decode(Payload.self, from: json),
+          payload.securityToken == UserScriptManager.securityTokenString else {
       return
     }
 
-    handleBraveRequestAdsEnabled()
+    switch payload.kind {
+    case .braveRequestAdsEnabled:
+      handleBraveRequestAdsEnabled(replyHandler)
+    case .launchNativeBraveTalk(let url):
+      guard let components = URLComponents(string: url),
+            case let room = String(components.path.dropFirst(1)),
+            let jwt = components.queryItems?.first(where: { $0.name == "jwt" })?.value
+      else {
+        return
+      }
+      launchNativeBraveTalk(room, jwt)
+      // TODO: Determine if we want to stop the load, close the tab, etc.
+      tab?.webView?.stopLoading()
```

Comment:
I'm wondering if `stopLoading()` would affect Release users. Maybe it would be better to move `guard Self.isIntegrationEnabled` check to the start of `userContentController` function. Wdyt?

Reply:
Yep your right, missed that, will fix